### PR TITLE
[SPARK-36020][SQL] Check logical link in remove redundant projects

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

The `RemoveRedundantProjects` feature can conflict with the AQE broadcast threshold ([PR](https://github.com/apache/spark/pull/32391)) sometimes. After removing the project, the physical plan to logical plan link can be changed and we may have a `Project` above `LogicalQueryStage`. This breaks AQE broadcast threshold, because the stats of `Project` does not have the `isRuntime = true` flag, and thus still use the normal broadcast threshold.

This PR updates `RemoveRedundantProjects` to not remove `ProjectExec` that has a different logical plan link than its child.

### Why are the changes needed?

Make AQE broadcast threshold work in more cases.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @allisonwang-db @ulysses-you @viirya

**Review by ulysses-you — COMMENTED** [review]
I think it's not related to AQE broadcast threshold. I verified this query in 3.1.2 and failed with
```
java.lang.IllegalStateException: UnspecifiedDistribution does not have default partitioning.
  at org.apache.spark.sql.catalyst.plans.physical.UnspecifiedDistribution$.createPartitioning(partitioning.scala:52)
  at org.apache.spark.sql.execution.exchange.EnsureRequirements$.$anonfun$ensureDistributionAndOrdering$1(EnsureRequirements.scala:54)
  at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at scala.collection.TraversableLike.map(TraversableLike.scala:238)
  at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
  at scala.collection.immutable.List.map(List.scala:298)
  at org.apache.spark.sql.execution.exchange.EnsureRequirements$.org$apache$spark$sql$execution$exchange$EnsureRequirements$$ensureDistributionAndOrdering(EnsureRequirements.scala:46)
  at org.apache.spark.sql.execution.exchange.EnsureRequirements$$anonfun$apply$1.applyOrElse(EnsureRequirements.scala:258)
  at org.apache.spark.sql.execution.exchange.EnsureRequirements$$anonfun$apply$1.applyOrElse(EnsureRequirements.scala:250)
```

The reason is the `ProjectExec` removed after `initialPlan`, but the logicalplan still has the `Project` node. Then during `replaceWithQueryStagesInLogicalPlan`, the physcialplan can not map to the origin `Porject` (due to `ProjectExec` removed ) so return the wrong logicalplan

```sql
Project [key#2L, c#140L, key#142L, c#141L]
+- Join Inner, (key#2L = key#142L)
   :- Project [c#140L, key#2L]
   :  +- Generate explode(c#5), [1], false, [c#140L]
   :     +- Project [(id#0L % 10) AS key#2L, array(id#0L, (id#0L + 1), (id#0L + 3)) AS c#5]
   :        +- Filter ((cast((id#0L * 2) as int) > 1) AND ((size(array(id#0L, (id#0L + 1), (id#0L + 3)), true) > 0) AND isnotnull((id#0L % 10))))
   :           +- Range (0, 10, step=1, splits=Some(8))
   +- Project [c#141L, key#142L]   <===  this Project should be wrapped inside  LogicalQueryStage
      +- LogicalQueryStage Generate explode(c#145), [1], false, [c#141L], BroadcastQueryStage 0 
```

At the end, `LogicalQueryStageStrategy` can not replace back to the exists broadcast join which has already planned in normal planner. That said the `BroadcastHashJoin` would be rePlanner during AQE which is unexpected.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala [hunk 1]
```diff
@@ -48,7 +48,7 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
   private def removeProject(plan: SparkPlan, requireOrdering: Boolean): SparkPlan = {
     plan match {
       case p @ ProjectExec(_, child) =>
-        if (isRedundant(p, child, requireOrdering)) {
+        if (isRedundant(p, child, requireOrdering) && canRemove(p, child)) {
           val newPlan = removeProject(child, requireOrdering)
           newPlan.setLogicalLink(child.logicalLink.get)
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala:53]: how about use

`newPlan.setLogicalLink(child.logicalLink.get)` -> `newPlan.setLogicalLink(p.logicalLink.get)` ?

**Comment by cloud-fan** [comment]
The root cause is the wrong mapping between physical and logical plans, and in 3.1 it causes query failures. My fix is to be safe and do not remove `ProjectExec` that may break the mapping.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala:53]: This looks risky. It makes it hard to reason about the mapping between physical and logical plans.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala:53]: I think we should even remove this line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala [hunk 2]
```diff
@@ -48,9 +48,8 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
   private def removeProject(plan: SparkPlan, requireOrdering: Boolean): SparkPlan = {
     plan match {
       case p @ ProjectExec(_, child) =>
-        if (isRedundant(p, child, requireOrdering)) {
+        if (isRedundant(p, child, requireOrdering) && canRemove(p, child)) {
           val newPlan = removeProject(child, requireOrdering)
-          newPlan.setLogicalLink(child.logicalLink.get)
           newPlan
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala:53]: just `removeProject(child, requireOrdering)`

**Review by ulysses-you — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/3.2

**Comment by cloud-fan** [comment]
@ulysses-you since 3.1 does not enable AQE by default, I didn't backport this PR. If you think it's necessary, feel free to open a backport PR, probably with a modified test case to show the problem in 3.1.

**Comment by ulysses-you** [comment]
@cloud-fan created https://github.com/apache/spark/pull/33238

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 3]
```diff
@@ -1929,6 +1929,29 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+
+  test("SPARK-36020: Check logical link in remove redundant projects") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1933]: For some reasons, this test is flaky or fails with Scala 2.13.
I filed a JIRA at https://issues.apache.org/jira/browse/SPARK-39525

**Pull request closed without merging**